src/methods/polygonize.jl

```diff
@@ -57,165 +51,278 @@ The implementation follows:
+    values=Base.union(A),
+    kw...
+)
+    # Create one feature per value
+    features = map(values) do value
+        multipolygon = polygonize(==(value), xs, ys, A)
+        GI.Feature(multipolygon; properties=(; value))
+    end 
+
+    return GI.FeatureCollection(features)
+end
+function polygonize(f::Base.Callable, xs::AbstractRange, ys::AbstractRange, A::AbstractMatrix; 
+    kw...
+)
+    # Make vectors of pixel bounds
+    xhalf = step(xs) / 2
+    yhalf = step(ys) / 2
+    # Make bounds ranges first to avoid floating point error making gaps or overlaps
+    xbounds = first(xs) - xhalf : step(xs) : last(xs) + xhalf
+    ybounds = first(ys) - yhalf : step(ys) : last(ys) + yhalf
+    Tx = eltype(xbounds)
+    Ty = eltype(ybounds)
+    xvec = Vector{Tuple{Tx,Tx}}(undef, length(xs))
+    yvec = Vector{Tuple{Ty,Ty}}(undef, length(ys))
+    for i in eachindex(xvec)
+        xvec[i] = xbounds[i], xbounds[i+1]
     end
-    ## If we filter off the minimum points, then it's a hair more efficient
-    ## not to convert contours with length < missingpoints to polygons.
-    if minpoints > 1
-        contours = Iterators.filter(contours) do contour
-            length(contour) > minpoints
-        end
-       return map(Polygon, contours)
-    else
-        return map(Polygon, contours)
+    for i in eachindex(yvec)
+        yvec[i] = ybounds[i], ybounds[i+1]
     end
+    return polygonize(f, xvec, yvec, A; kw...)
 end
 
-## rotate direction clockwise
-rot_clockwise(dir) = (dir) % 8 + 1
-## rotate direction counterclockwise
-rot_counterclockwise(dir) = (dir + 6) % 8 + 1
-
-## move from current pixel to next in given direction
-function move(pixel, image, dir, dir_delta)
-    newp = pixel + dir_delta[dir]
-    height, width = size(image)
-    if (0 < newp[1] <= height) && (0 < newp[2] <= width)
-        if image[newp] != 0
-            return newp
-        end
+function updateval(dict, key, val)
+    if haskey(dict, key)
+        existingval = dict[key][1][1]
+        newval = ((existingval, val), (true, true))
+        dict[key] = newval 
+    else
+        newval = ((val, val), (true, false))
+        dict[key] = newval 
     end
-    return CartesianIndex(0, 0)
-end
-
-## finds direction between two given pixels
-function from_to(from, to, dir_delta)
-    delta = to - from
-    return findall(x -> x == delta, dir_delta)[1]
 end
 
-function detect_move(image, p0, p2, nbd, border, done, dir_delta)
-    dir = from_to(p0, p2, dir_delta)
-    moved = rot_clockwise(dir)
-    p1 = CartesianIndex(0, 0)
-    while moved != dir ## 3.1
-        newp = move(p0, image, moved, dir_delta)
-        if newp[1] != 0
-            p1 = newp
-            break
+function polygonize(f, xs::AbstractVector{T}, ys::AbstractVector{T}, A::AbstractMatrix; 
+    minpoints=0,
+) where T
+    # Define buffers for edges and rings
+    edges = Dict{T,Tuple{Tuple{T,T},Tuple{Bool,Bool}}}()
+    rings = Vector{T}[]
+
+    @assert (length(xs), length(ys)) == size(A)
+
```

Comment:
Maybe less skipped lines here for vertical compression?

Reply:
I need a proper error rather than `@assert toolll